### ISSUE TITLE
using custom attention mask

### ISSUE THREAD

**Issue by whatisslove11** [Dao-AILab/flash-attention#584]
Hello!
I am doing a translation task and would like to try using flash attention in my model
In addition to the usual triangular mask, I also need to mask padding tokens so that the model does not pay attention to them - sequences of the same length already arrive in the model itself
As I understand it, there is no function of feeding your mask yet
Could you tell me how I can use my mask or make flash attention add padding tokens to the mask itself?

**Comment by whatisslove11** [comment]
@tridao

**Comment by tridao** [comment]
You can look at how we do it in [BERT](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/models/bert.py): Remove all padding tokens before the first layer.

Idk if that works for translation.

**Comment by flower-with-safe** [comment]
what if I want to use reset-attention-mask when pretrain a llama model? 
for example, my attention mask could be:
tensor([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
             [1., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
             [1., 1., 1., 0., 0., 0., 0., 0., 0., 0.],
             [1., 1., 1., 1., 0., 0., 0., 0., 0., 0.],
             [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
             [0., 0., 0., 0., 1., 1., 0., 0., 0., 0.],
             [0., 0., 0., 0., 1., 1., 1., 0., 0., 0.],
             [0., 0., 0., 0., 1., 1., 1., 1., 0., 0.],
             [0., 0., 0., 0., 1., 1., 1., 1., 1., 0.],
             [0., 0., 0., 0., 1., 1., 1., 1., 1., 1.]])
in such case how could I use flash attention? 
@tridao

**Comment by sentialx** [comment]
@tridao Is it possible to use a PrefixLM attention mask?

**Comment by tridao** [comment]
No that's not supported.

**Comment by sentialx** [comment]
It looks like custom attention mask is the most requested feature. In fact, it seems to be already solved by this PR: https://github.com/Dao-AILab/flash-attention/pull/57

**Comment by iiLaurens** [comment]
Am also much interested in custom masks. I think the value of Prefixlm mask is not appreciated enough. Would like to experiment continuation of pretraining with prefixlm  a la UL2R[UL2R](https://blog.research.google/2022/11/better-language-models-without-massive.html?m=1).

@tridao, are you considering support for a causal mask input? Or do you have specific objections to it?
